build: introduce dynamo base container

nv-tusharma opened the pull request:
#### Overview:

Milestone 1 in DEP: 
https://github.com/ai-dynamo/enhancements/pull/8

This PR adds support for a Dynamo base container which can be used for faster prototyping and faster developer validation for PR changes. As part of the PR, we change the pre-merge check to build this Dynamo base image instead of the vLLM full fat container. As a result of this change and further optimizations to the build, we are able to reduce the time to run PR validation from 1 hour approximately to 20 minutes. 

#### Details:

- Introduce a Dockerfile (Dockerfile.dynamo) which provides NIXL/UCX, Dynamo, and NATS/ETCD
- Change the github pre-merge-workflow to build the Dynamo base container instead of the full fat vLLM container
- Remove Dockerfile.none and add Dockerfile.dynamo as a replacement
- Add integration testing as a pre-merge check for validation, use docker compose to start NATS & ETCD services

#### Where should the reviewer start?

- Dockerfile.dynamo

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

ref: OPS-565, OPS-607

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Introduced a new multi-stage Dockerfile for the Dynamo project, supporting multiple architectures and enhanced build processes.
* **Chores**
  * Updated the build script and Docker-related configurations to replace the "NONE" framework with "DYNAMO".
  * Updated the .dockerignore file to exclude certain Dockerfiles from the build context.
  * Removed the old Dockerfile for the "NONE" framework.
  * Enhanced the GitHub Actions workflow to support the "dynamo" framework, expanded testing, and improved service management during CI.
* **Tests**
  * Marked a failing integration test to be ignored during CI runs.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nv-tusharma committed:
build: introduce dynamo base container and remove current Dockerfile.none

nv-tusharma committed:
minor dockerfile fixes + use github dockerfile in premerge

nv-tusharma committed:
Use dev target for dynamo build

nv-tusharma committed:
Add NATS + ETCD installation to dynamo base

nv-anants committed:
remove rust build

nv-anants committed:
Add back nats/etcd

nv-anants committed:
Add back protobuf

nv-tusharma committed:
Remove Dockerfile.none from /containers

nv-anants committed:
add nixl

nv-tusharma committed:
Install essential python dependencies for pytest testing

nv-anants committed:
move installs to last

nv-anants committed:
update

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

nv-anants committed:
Revert back to whl install

nv-anants committed:
improvements

nv-anants committed:
ADd integration tests

nv-anants committed:
docker compose

nv-anants committed:
add tests

nv-tusharma commented on .github/workflows/pre-merge-python.yml:
```diff
@@ -46,10 +46,21 @@ jobs:
-          docker run -w /workspace/lib/llm --name ${{ env.CONTAINER_ID }}_rust_checks ${{ steps.define_image_tag.outputs.image_tag }} bash -ec 'rustup component add rustfmt clippy && cargo fmt -- --check && cargo clippy --features block-manager --no-deps --all-targets -- -D warnings && cargo test --locked --all-targets --features=block-manager'
+          docker compose up -d nats-server etcd-server
+          # Wait for services to be ready
+          sleep 15
```
Do we need this sleep command? Can we make the compose operation blocking in some way?

nv-anants committed:
skip

nv-anants committed:
Merge branch 'main' into dynamo-base-container-dev

nv-tusharma committed:
Remove sleep 15 to test if docker compose is a blocking op

rmccorm4 commented on .github/workflows/pre-merge-python.yml:
```diff
@@ -46,10 +46,19 @@ jobs:
+          docker compose up -d nats-server etcd-server
+      - name: Run Rust checks (block-manager + integration tests)
+        run: |
+          docker run -w /workspace/lib/llm --name ${{ env.CONTAINER_ID }}_rust_checks ${{ steps.define_image_tag.outputs.image_tag }} bash -ec 'rustup component add rustfmt clippy && cargo fmt -- --check && cargo clippy --features block-manager --no-deps --all-targets -- -D warnings && cargo test --locked --all-targets --features=block-manager && cargo test --locked --features integration -- --no ...
```
What's the general plan here on testing rust things inside dynamo container (pre-merge-python [should be renamed]) and vs outside container (pre-merge-rust)?

My understanding was that only the KVBM tests were put in here because `pip install nixl` didn't have the necessary NIXL shared libraries to test it outside of the container.

For everything else on the rust test side, it's pretty lightweight and runs in a vanilla ubuntu container with nothing but cargo: https://github.com/ai-dynamo/dynamo/blob/faafa5ffe67c154597e8c68fbc565d68aedb2da2/.github/workflows/pre-merge-rust.yml#L39

mc-nv commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+### UCX EFA Setup ###
+RUN rm -rf /opt/hpcx/ucx && \
+    rm -rf /usr/local/ucx && \
+    echo "Building UCX with reference $NIXL_UCX_REF" && \
+    cd /usr/local/src &&                            \
+    git clone --depth 1 --branch $NIXL_UCX_REF https://github.com/openucx/ucx.git && \
+    cd ucx &&                                       \
+    ./autogen.sh && ./configure                     \
+    --prefix=/usr/local/ucx                         \
+    --enable-shared                                 \
+    --disable-static                                \
+    --disable-doxygen-doc                           \
+    --enable-optimizations                          \
+    --enable-cma                                    \
+    --enable-devel-headers                          \
+    --with-cuda=/usr/local/cuda                     \
+    --with-verbs                                    \
+    --with-efa                                      \
+    --with-dm                                       \
+    --with-gdrcopy=/usr/local                       \
+    --enable-mt &&                                  \
+    make -j$(nproc) &&                              \
+    make -j$(nproc) install-strip &&                \
+    ldconfig && \
+    cd /usr/local/src && \
+    rm -rf ucx
```
```suggestion
    rm -rf /usr/local/ucx && \
    echo "Building UCX with reference $NIXL_UCX_REF" && \
    cd /usr/local/src && \
    git clone --depth 1 --branch $NIXL_UCX_REF https://github.com/openucx/ucx.git && \
    cd ucx && \
    ./autogen.sh && \
    ./configure \
      --prefix=/usr/local/ucx \
      --enable-shared \
      --disable-static \
      --disable-doxygen-doc \
      --enable-optimizations \
      --enable-cma \
      --enable-devel-headers \
      --with-cuda=/usr/local/cuda \
      --with-verbs \
      --with-efa \
      --with-dm \
      --with-gdrcopy=/usr/local \
      --enable-mt && \
    make -j$(nproc) &&                              \
    make -j$(nproc) install-strip &&                \
    ldconfig && \
    cd /usr/local/src && \
    rm -rf ucx
```

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+FROM ${BASE_IMAGE}:${BASE_IMAGE_TAG} AS base
+
+# Redeclare ARCH and ARCH_ALT so they're available in this stage
+ARG ARCH
+ARG ARCH_ALT
```
uname -m is definitely a better approach, it's what we did for Triton in the past. Will follow up in upcoming PRs

nv-tusharma replied:
I've created OPS-592 to track progress of this.

alec-flowers commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+    PATH="/opt/dynamo/venv/bin:${PATH}"
+
+##################################
+##### Wheel Build Image ##########
```
Can someone explain the philosophy of a wheel build image? I feel that we loose any caching ability in this stage. Lets say I make a one line rust change, everything gets fully recompiled, when in reality it should be a marginal diff in the compilation. 

Its also my understanding that binaries take up a lot of the rust compile time. Can we remove any cargo binaries?

nv-tusharma replied:
@nv-anants did some work to remove some of the cargo binaries that weren't required. The main purpose of the wheel build image stage is to ensure we build dynamo on a compatible glibc version that aligns with NIXL version. This is also the stage we use for picking up artifacts for release. There may be a path where we can seperate this logic into it's own build and Dockerfile.

alec-flowers commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+FROM quay.io/pypa/manylinux_2_28_${ARCH_ALT} AS wheel_builder
+
+ARG CARGO_BUILD_JOBS
```
Do we actually override CARGO_BUILD_JOBS and other things to speed up compilation? I would expect our build nodes to be able to handle many times 16 threads and we should be leveraging that. 

If we are doing more than 16 threads are we sure whatever number we chose gives us the fastest builds?

alec-flowers replied:
Ok, do we do it in github? What is it set too here?

nv-tusharma committed:
minor syntax fix for NIXL install

Co-authored-by: Misha Chornyi <99709299+mc-nv@users.noreply.github.com>
Signed-off-by: Tushar Sharma <tusharma@nvidia.com>

nv-tusharma committed:
Use dnf and minor optimizations in NIXL/UCX build to reduce steps

rmccorm4 commented on container/Dockerfile.dynamo:
In future do we plan to simplify `Dockerfile.{vllm,trtllm,sglang}` to just be the framework specific deltas?

ex:
```
# Dockerfile.vllm
FROM dynamo_base_image
RUN pip install vllm
...
```
and automate with `build.sh` so that under the hood it does something like:
```bash
docker build -f Dockerfile.dynamo
docker build -f Dockerfile.vllm # depends on previous step
```

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+# Install NIXL Python module
+# TODO: Move gds_path selection based on arch into NIXL build
+# TEMP: disable gds backend for arm64
```
Fixed in latest revision.

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+# Install Rust using RUSTARCH derived from ARCH_ALT
+RUN wget --tries=3 --waitretry=5 "https://static.rust-lang.org/rustup/archive/1.28.1/${RUSTARCH}/rustup-init" && \
+    # TODO: Add SHA check back based on RUSTARCH
```
Created OPS-591 to track this task.

rmccorm4 commented on container/Dockerfile.none:
Let's remove the NONE references from `run.sh` too

rmccorm4 replied:
Scratch that - they're unrelated.

nnshah1 replied:
hmm - but shouldn't there be changes in run.sh to support this framework? all build targets should have a corresponding run target ....

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+    else \
+        nixl_build_args=""; \
+    fi && \
+    mkdir build && \
```
Updated in latest PR.

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+RUN yum update -y \
+    && yum install -y llvm-toolset protobuf-compiler python3.12-devel \
+    && yum clean all \
+    && rm -rf /var/cache/yum
```
In latest commit, I migrated to use dnf instead of yum. basic sanity/integration testing passes as expected.

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

nv-tusharma committed:
Rename pre-merge-python.yml to build-and-test.yml for better clarity

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

biswapanda commented on container/build.sh:
```diff
@@ -406,8 +406,8 @@ if [[ $FRAMEWORK == "VLLM" ]]; then
     DOCKERFILE=${SOURCE_DIR}/Dockerfile.tensorrt_llm
-elif [[ $FRAMEWORK == "NONE" ]]; then
-    DOCKERFILE=${SOURCE_DIR}/Dockerfile.none
+elif [[ $FRAMEWORK == "DYNAMO" ]]; then
```
Thanks `dynamo` is a much better name for base framework-less Dockerfile

saturley-hall commented on .github/workflows/build-and-test.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
       matrix:
         framework:
-          - vllm
+          - dynamo
```
I am not in love with this unless we see a future in which there are going to be building many of these containers in parallel. That is typically what you use the `strategy.matrix` for.

saturley-hall commented on container/Dockerfile.none:
```diff
@@ -1,58 +0,0 @@
-FROM ubuntu:24.04 AS dev
```
Are we sure no one is using this?
@alec-flowers for viz

mc-nv commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+ARG BASE_IMAGE_TAG="25.01-cuda12.8-devel-ubuntu24.04"
+ARG RELEASE_BUILD
+
+# Define general architecture ARGs for supporting both x86 and aarch64 builds.
+#   ARCH: Used for package suffixes (e.g., amd64, arm64)
+#   ARCH_ALT: Used for Rust targets, manylinux suffix (e.g., x86_64, aarch64)
+#
+# Default values are for x86/amd64:
+#   --build-arg ARCH=amd64 --build-arg ARCH_ALT=x86_64
+#
+# For arm64/aarch64, build with:
+#   --build-arg ARCH=arm64 --build-arg ARCH_ALT=aarch64
+#
+# NOTE: There isn't an easy way to define one of these values based on the other value
+# without adding if statements everywhere, so just define both as ARGs for now.
```
This comment only increase file size

mc-nv commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+ARG NIXL_REF=0.4.1
+
+# Environment variables for NIXL
+ENV NIXL_SRC_DIR=/opt/nixl \
+    NIXL_PREFIX=/opt/nvidia/nvda_nixl \
+    NIXL_LIB_DIR=/opt/nvidia/nvda_nixl/lib/${ARCH_ALT}-linux-gnu \
+    NIXL_PLUGIN_DIR=/opt/nvidia/nvda_nixl/lib/${ARCH_ALT}-linux-gnu/plugins \
+    LD_LIBRARY_PATH=/opt/nvidia/nvda_nixl/lib/${ARCH_ALT}-linux-gnu:/opt/nvidia/nvda_nixl/lib/${ARCH_ALT}-linux-gnu/plugins:$LD_LIBRARY_PATH
```
It's better to hard code it, unless you plan to override it with build args

nv-tusharma committed:
Revert changes to sglang and trt llm dockerfiles

saturley-hall commented on container/build.sh:
```diff
@@ -49,7 +49,7 @@ PYTHON_PACKAGE_VERSION=${current_tag:-$latest_tag.dev+$commit_id}
 # installed within framework specific sections of the Dockerfile.
 
-declare -A FRAMEWORKS=(["VLLM"]=1 ["TENSORRTLLM"]=2 ["NONE"]=3 ["SGLANG"]=4)
+declare -A FRAMEWORKS=(["VLLM"]=1 ["TENSORRTLLM"]=2 ["DYNAMO"]=3 ["SGLANG"]=4)
```
I am slightly uncomfortable with calling this a framework, to me that implies that we have a backend that is capable of running LLMs.

What makes sense to me is less clear. @nnshah1 do you have an opinion around the ergonomics here?

nv-tusharma replied:
I won't be able to address this in this PR since this will require major refactoring of the build.sh script to support dynamo-component specific installation. For now, I've created OPS-613 to track the work for this request.

nnshah1 replied:
I would hold off on adding more targets and more complexity here - I think we want a few containers and not try to mirror each component - I think that is more appropriate for the pip install path - container builds should have a few as needed 

A few thoughts

1) Should this be a target or base container that is used for all frameworks? (is that the goal - I need to review the DEP again)

2) I agree with @saturley-hall here we should be a little careful in expectation - "dynamo" as framework almost suggest it is a superset and not a subset. 

What if we kept the framework as "NONE" and remove the suffix on the Dockerfile?

So we would have:

Dockerfile
Dockerfile.vllm
Dockerfile.sglang
...

I think "None" works for a framework name - but "Dockerfile.none" is odd -

nv-tusharma replied:
For now, I've set the framework to none in build.sh and changed the Dockerfile name from Dockerfile.dynamo to Dockerfile.  I think this is ok for now.

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+    uv pip install /workspace/wheelhouse/ai_dynamo*any.whl && \
+    uv pip install /workspace/wheelhouse/nixl/nixl*.whl
+
+RUN uv pip install /workspace/benchmarks
```
Is there a reason to do this in a separate layer? Moreover do we think this code will churn more or less frequently than the dynamo code? If it is less frequently it should come first for layer caching? If more frequently then it is fine where it is. If the same amount I would want it in the previous run command to minimize layers

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+COPY . /workspace
+
+# Common dependencies
+RUN --mount=type=bind,source=./container/deps/requirements.txt,target=/tmp/requirements.txt \
```
You just copied in the `workspace` in the previous layer, I don't think you need to bind here. However I would say that you will want to move this _before_ you copy in the workspace otherwise with every change to the codebase you will be reinstalling the same requirements regardless of if they have changed

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+RUN --mount=type=bind,source=./container/deps/requirements.test.txt,target=/tmp/requirements.txt \
+    uv pip install --requirement /tmp/requirements.txt
+
+RUN uv pip install /workspace/wheelhouse/ai_dynamo_runtime*cp312*.whl && \
```
Do you want to keep the wheels around after you copy them in?
I mean I guess probably you do I just want you to think about it.

nv-tusharma replied:
Yes, since we can then use these wheels in other backend-specific builds as wheels that can be installed in the runtime container.

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+# for details and reproducer to manually test if the image
+# can be updated to later versions.
+ARG BASE_IMAGE_TAG="25.01-cuda12.8-devel-ubuntu24.04"
+ARG RELEASE_BUILD
```
Default this to false?

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+    rm rustup-init && \
+    chmod -R a+w $RUSTUP_HOME $CARGO_HOME
+
+RUN apt-get update -y && \
```
I would like to see some consideration of the ordering here: alphabetical, by the thing it is a dependency for, or similar organizing logic.

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+    uv pip install --requirement /tmp/requirements.txt
+
+# Install test dependencies
+RUN --mount=type=bind,source=./container/deps/requirements.test.txt,target=/tmp/requirements.txt \
```
Same as the other dependencies, we should move these higher in the stage if they are not going to be changing frequently

saturley-hall reviewed (changes requested):
Overall I like the direction.

I think there are some opportunities for changes that will be positive in terms of caching for developers doing local builds or our CI once caching becomes possible.

I would like @nnshah1 to weigh in on ergonomics since he is much closer to the naming standards that we are trying to arrive on.

nv-tusharma committed:
nit:address minor comments + move pip install commands earlier to improve docker layer caching perf

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

nv-tusharma committed:
Install benchmarks in the same layer as dynamo installation

nv-tusharma committed:
Revert "Merge branch 'main' into dynamo-base-container-dev"

This reverts commit ba701cd0be05d010f587ff6058fa0d444384273d, reversing
changes made to f735136b7552e0e4600eb918304d502b2e255439.

nv-tusharma committed:
Reapply "Merge branch 'main' into dynamo-base-container-dev"

This reverts commit bc9d37655ab66d01bba94b9ec09991e38e1e824c.

grahamking commented:
I am very much looking forward to faster image builds. Thanks for the work you are doing here.

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

ryanolson commented:
Great start. Are we building this base image every time? If so are we getting good cache reuse if the dockerfile hasn't changed?

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+    chmod -R a+w $RUSTUP_HOME $CARGO_HOME
+
+RUN apt-get update -y && \
+    apt-get install -y \
```
```suggestion
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
```
Unsure of necessity here but I would consider this as a best practice as [taken from here](https://github.com/ai-dynamo/dynamo/blob/ccc8815b35003168a8b56295a776d5860e5dcebf/container/Dockerfile.vllm#L439C5-L439C80)

@whoisj would be the expert on that here

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+# Common dependencies
+RUN --mount=type=bind,source=./container/deps/requirements.txt,target=/tmp/requirements.txt \
+uv pip install --requirement /tmp/requirements.txt
```
```suggestion
    uv pip install --requirement /tmp/requirements.txt
```

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+# Install test dependencies
+RUN --mount=type=bind,source=./container/deps/requirements.test.txt,target=/tmp/requirements.txt \
+uv pip install --requirement /tmp/requirements.txt
```
```suggestion
    uv pip install --requirement /tmp/requirements.txt
```

saturley-hall commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+uv pip install --requirement /tmp/requirements.txt
+
+# Install test dependencies
+RUN --mount=type=bind,source=./container/deps/requirements.test.txt,target=/tmp/requirements.txt \
```
Any reason not to combine this with the lines above into a singe `RUN` layer?

saturley-hall reviewed (changes requested):
Minor formatting changes.

The `apt-get` and layer combination questions I leave to your judgement. If you have questions about the former I would reach out to @whoisj

nnshah1 commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+# Please check https://github.com/ai-dynamo/dynamo/pull/1065
+# for details and reproducer to manually test if the image
+# can be updated to later versions.
+ARG BASE_IMAGE_TAG="25.01-cuda12.8-devel-ubuntu24.04"
```
for future -is this common accross frameworks? if so should we move the base image into build.sh as arg there?

nv-tusharma replied:
TRT-LLM is the only container which uses a separate base container, I also suspect that there will be discrepancies across CUDA versions used by a particular framework (including for performance).

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

nv-tusharma committed:
nit: introduce minor improvements and suggestions

nv-tusharma committed:
Change framework CLI arg to none

nv-tusharma committed:
Mount workspace as a volume for rust + pytest checks

nv-tusharma committed:
Pass in is_static:False into entrypoint integration test

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+
+WORKDIR /workspace
+
+COPY --from=wheel_builder /workspace/dist/*.whl /workspace/wheelhouse/
```
Updated in latest revision to use /opt/dynamo as the WORKDIR for hosting artifacts instead of WORKSPACE.

nv-tusharma commented:
Adding Misha's comments here for transparency:
1. Makes sure that dependency and installation has separate stage: The dockerfile leverages a multi-stage approach with 3 primary stages: base, wheel_builder, and dev. Base builds all the dynamo dependencies (NIXL, UCX, python deps), wheelbuilder builds the dynamo wheel, and dev installs the artifacts built from previous stages onto the container. 
2. Library configuration should be resolved in ld.so.conf: Removed references to LD_LIBRARY_PATH and used ld.so.conf alternatively. 
3. Final directory structure shouldn't contain directory like /workspace it's a bad practice. Changed to host artifacts in /opt/dynamo instead of /workspace. 
4. Comments within single instruction step must be removed or moved before the step. I've kept this for now since it improves clarity of which apt dependencies map to the package to install. 
5. python virtual environment, may be if it works fine we can compile minimal version of python with pip and use it by having full control of it: See L171-177, we are using a python virtual environment and installing all python-specific packages to that location.  
6. NIXL build, wondering if build configuration could handle the logic you place in Dockerfile. OPS-597 to handle this task.

nv-tusharma committed:
nit:minor formatting fix

nv-tusharma committed:
Merge branch 'main' into dynamo-base-container-dev

ryanolson commented:
> @ryanolson 
> 
> > Great start. Are we building this base image every time? If so are we getting good cache reuse if the dockerfile hasn't changed?
> 
> 
> 
> We don't have caching setup for containers built via Github due to issues with the container registry. As next steps, we will look to setup some external registry which we can use to enable external caching. For now, the Dockerfile is setup to use docker's default build cache mechanism which keeps instructions which are expected to change more frequently (such as dynamo src change) towards the end of the Dockerfile instead of the beginning. 

Can we use NGC?

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,268 @@
+#
+# For arm64/aarch64, build with:
+#   --build-arg ARCH=arm64 --build-arg ARCH_ALT=aarch64
+#TODO OPS-592: Leverage uname -m to determine ARCH instead of passing it as an arg
```
I think by default we should use uname -m as the default mechanism for building and override as necessary by passing in a different architecture as a build-arg?

nv-tusharma commented on container/Dockerfile.dynamo:
```diff
@@ -0,0 +1,267 @@
+RUN dnf update -y \
+    && dnf install -y llvm-toolset protobuf-compiler python3.12-devel \
+    && dnf clean all \
+    && rm -rf /var/cache/dnf
```
Will follow up in upcoming PRs

nv-tusharma commented:
@ryanolson NGC would definitely be a viable option.
